pkg/agent/ipassigner/linkdetector/link_detector_linux_test.go

```diff
@@ -0,0 +1,230 @@
+				newLinkEvent(false, "eth1", 3),
+			},
+			expectNetlinkCalls: func(nl *netlinktesting.MockInterfaceMockRecorder) {
+				nl.LinkList().Return([]netlink.Link{
+					newLink("lo", 1),
+					newLink("eth0", 2),
+				}, nil)
+			},
+		},
+		{
+			name: "watch eth1",
+			eventHandlers: []*linkEventHandler{
+				newLinkEventHandler([]string{
+					"eth1",
+				}, "eth1"),
+			},
+			linkEvents: []netlink.LinkUpdate{
+				newLinkEvent(false, "eth1", 3),
+			},
+			expectNetlinkCalls: func(nl *netlinktesting.MockInterfaceMockRecorder) {
+				nl.LinkList().Return([]netlink.Link{
+					newLink("lo", 1),
+					newLink("eth0", 2),
+				}, nil)
+			},
+		},
+		{
+			name: "watch eth1 and eth1 renamed",
+			eventHandlers: []*linkEventHandler{
+				newLinkEventHandler([]string{
+					"eth1", // initial notification
+					"eth1", // rename notification
+				}, "eth1"),
+			},
+			linkEvents: []netlink.LinkUpdate{
+				newLinkEvent(false, "eth1~", 3),
+			},
+			expectNetlinkCalls: func(nl *netlinktesting.MockInterfaceMockRecorder) {
+				nl.LinkList().Return([]netlink.Link{
+					newLink("lo", 1),
+					newLink("eth0", 2),
+					newLink("eth1", 3),
+				}, nil)
+			},
+		},
+		{
+			name: "watch eth1, eth1 renamed and created with new index",
+			eventHandlers: []*linkEventHandler{
+				newLinkEventHandler([]string{
+					"eth1", // initial notification
+					"eth1", // rename notification
+					"eth1", // new link notification
+				}, "eth1"),
+			},
+			linkEvents: []netlink.LinkUpdate{
+				newLinkEvent(false, "eth1~", 3),
+				newLinkEvent(false, "eth1", 4),
+			},
+			expectNetlinkCalls: func(nl *netlinktesting.MockInterfaceMockRecorder) {
+				nl.LinkList().Return([]netlink.Link{
+					newLink("lo", 1),
+					newLink("eth0", 2),
+					newLink("eth1", 3),
+				}, nil)
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			ctrl := gomock.NewController(t)
+			defer ctrl.Finish()
+
+			prevLinkSubscribeFunc := linkSubscribeFunc
+			defer func() { linkSubscribeFunc = prevLinkSubscribeFunc }()
+			linkSubscribeFunc = func(ch chan<- netlink.LinkUpdate, done <-chan struct{}, options netlink.LinkSubscribeOptions) error {
+				go func() {
+					for _, e := range tt.linkEvents {
+						ch <- e
+					}
+				}()
+				return nil
+			}
+
+			netlink := netlinktesting.NewMockInterface(ctrl)
+			tt.expectNetlinkCalls(netlink.EXPECT())
+
+			d := &linkDetector{
+				eventHandlers: map[string][]LinkEventHandler{},
+				linkNames:     set.New[string](),
+				linkIndexMap:  map[int32]string{},
+				netlink:       netlink,
+			}
+			for _, h := range tt.eventHandlers {
+				d.AddEventHandler(h.onLinkEvent, h.watchLinkNames...)
+			}
+
+			stopCh := make(chan struct{})
+			defer close(stopCh)
+			go d.listAndWatchLinks(stopCh)
+			assert.Eventuallyf(t,
+				func() bool {
+					for _, l := range tt.eventHandlers {
+						l.lock.Lock()
+						defer l.lock.Unlock()
+						if !reflect.DeepEqual(l.expectedLinkEvents, l.receivedEvents) {
```

Comment:
use `assert.EventuallyWithT` and use a testify assertion instead of `reflect.DeepEqual` directly
you can then also remove the `t.Logf` call